aesara/link/jax/dispatch/random.py

```diff
@@ -326,3 +326,34 @@ def sample_fn(rng, size, dtype, *parameters):
         return (rng, sample_exp)
 
     return sample_fn
+
+
+@jax_sample_fn.register(aer.MultinomialRV)
+def jax_sample_fn_multinomial(op):
+    """JAX implementation of `MultinomialRV`."""
+
+    def _categorical(key, p, shape):
+        shape = shape or p.shape[:-1]
+        s = jax.numpy.cumsum(p, axis=-1)
+        r = jax.random.uniform(key, shape=shape + (1,))
+
+        return jax.numpy.sum(s < r, axis=-1)
```

Comment:
Is that different from `jax.random.categorical`?

Reply:
Yes, the `jax.random.categorical` uses their implementation of the Gumbel distribution. However, after using `git blame`, it seems the NumPyro team used this implementation before the implementation of the `jax.random.categorical`. 

If we would like to update this code to use the `jax.random.categorical`, it would be the following:

```python3

def sample_fn(rng, size, dtype, *parameters):

        rng_key = rng["jax_state"]
        rng_key, sampling_key = jax.random.split(rng_key, 2)

        n, p = parameters
        n_max = jax.numpy.max(n)
        size = size or p.shape[:-1]
    
        logits = jax.scipy.special.logit(p)
        indices = jax.random.categorical(jax_key, logits, shape=(n_max,) + size)
        one_hot = jax.nn.one_hot(indices, p.shape[0])
        sample = jax.numpy.sum(one_hot, axis=0, dtype=dtype, keepdims=False)

        rng["jax_state"] = rng_key

        return (rng, sample)
```

Reply:
That looks much simpler, great!

Since the multinomial distribution is slightly more complex than the other ones when it comes to shapes we should make sure that the output shape of the samples that are generated in the JAX backend is identical to those of the samples generated with the other backends.

Reply:
I tend to use `jax.random.choice` for this kind of thing. jax.random.categorical (via Gumbel) has a quadratic complexity.

![image](https://user-images.githubusercontent.com/19948263/209123329-c909174d-6968-4478-88f1-e7c6a186f475.png)

Reply:
You just need to check the Gumbel implementation, they form a N by N array.